libtmux/test.py

```diff
@@ -34,13 +34,19 @@ def retry(seconds=RETRY_TIMEOUT_SECONDS):
     Examples
     --------
 
-    >>> while retry():
+    >>> def f():
     ...     p = w.attached_pane
     ...     p.server._update_panes()
-    ...     if p.current_path == pane_path:
-    ...         break
+    ...     return p.current_path == pane_path
+    ...
+    ... retry(f)
     """
-    return (lambda: time.time() < time.time() + seconds)()
+    ini = time.time()
+
+    while not fun():
+        end = time.time()
+        if end - ini >= seconds:
+            raise Exception('Function timed out without meeting condition')
```

Comment:
I feel a bit resistant of moving it to a callback. It may be the only choice.

We have many `while retry()` in `test_workspacebuilder.py` ([link](https://github.com/tmux-python/tmuxp/blob/b0e744ec4d8ae83bf877f2c62a3334773970828f/tests/test_workspacebuilder.py)).  This turns the `while` block of code into a callback.

I'd like to see what other options are out there so we're not using callbacks, if it's reasonable to avoid it.

Any chance you could also find a `with` compatible one? e.g. [interruptingcow](https://bitbucket.org/evzijst/interruptingcow/src/master/) ([src](https://bitbucket.org/evzijst/interruptingcow/src/master/interruptingcow/__init__.py))

I'm happy to keep this function as well (in case we want to use this style of invocation or ultimately move to it)

Reply:
> I gave it some thought in terms of a context manager, but honestly I don't think it is a good idea to add that much complexity (and its [own bugs](https://bitbucket.org/evzijst/interruptingcow/issues?status=new&status=open)) to such a simple use case. I do think spawning a thread/task/context/whatever and reentrantly time it out is actually a bad idea that will complicate debugging.

I think the solution we've come to makes the most sense. I've seen solutions like interruptingcow's on stackoverflow and found the lack of compatibility also a bit troubling.

Would be awesome of python standard library had cross-platform, sync + async, bullet proof solution for this. But it may not be that easy to solve :)

> I have to say that my proposed code is intended to replace the current testing-only retry() function that is broken, but not intended for using in user code like plugins, or tmuxp code itself.

This also is fine with me.